File: backend/app/api/routes/approvals.py
"""approvals.py — 写操作工具审批管理 API

供管理员（或运维后台）查看/批准/驳回审批单。
鉴权: 走全局 api_key_middleware（X-API-Key），与其他管理路由一致。

P3（docs/auth 03 修复清单②）：reviewer 不再默认 "admin"——
请求体可传 reviewer 覆盖，但缺省时取网关注入的 X-User-Name；
都拿不到记 "unknown" 而不是冒充 admin（审计字段不许有默认身份）。
"""
from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from backend.app.api.deps import OperatorIdentity, resolve_operator_role
from backend.app.api.identity import resolve_identity
from backend.security.tool_approval import decide_request, is_degraded, list_requests

router = APIRouter(prefix="/approvals", tags=["工具审批"])


class ApprovalDecision(BaseModel):
    reviewer: str = Field("", description="审批人标识；缺省取网关身份头（X-User-Name/X-User-Id）")
    reason: str = Field("", description="批准/驳回理由")


def _resolve_reviewer(request: Request, body: ApprovalDecision,
                      operator: OperatorIdentity | None = None) -> str:
    if body.reviewer.strip():
        return body.reviewer.strip()
    # JWT 通道优先用 operator 审计标识（user:<id>）；服务凭据通道回退身份头
    if operator is not None and operator.actor.startswith("user:"):
        return operator.actor
    ident = resolve_identity(request)
    return ident.user_name or ident.user_id or "unknown"


def _require_admin(operator: OperatorIdentity) -> None:
    """批准/驳回是写操作处置权，映射到 prompts RBAC 的最高档：仅 admin。

    viewer/editor 在管理端审批页能看到队列（read 不限），但处置按钮会收到 403。
    服务凭据（internal-token）映射 admin，内部闭环不受影响。
    """
    if operator.role != "admin":
        raise HTTPException(403, f"仅 admin 可处置审批单（当前角色 {operator.role}）")


@router.get("")
async def list_approvals(status: str = "", limit: int = 50):
    """审批单列表。status=pending/approved/rejected/executed，空=全部。"""
    limit = max(1, min(limit, 200))
    return {
        "degraded": is_degraded(),
        "items": list_requests(status=status or None, limit=limit),
    }


@router.post("/{request_id}/approve")
async def approve(request_id: str, body: ApprovalDecision, request: Request,
                  operator: OperatorIdentity = Depends(resolve_operator_role)):
    """批准审批单。批准后 TTL 内重试相同操作即可执行（仅 admin）。"""
    _require_admin(operator)
    rec = decide_request(request_id, approve=True,
                         reviewer=_resolve_reviewer(request, body, operator), reason=body.reason)
    if rec is None:
        raise HTTPException(404, "审批单不存在或已处理")
    return {"ok": True, "request": rec}


@router.post("/{request_id}/reject")
async def reject(request_id: str, body: ApprovalDecision, request: Request,
                 operator: OperatorIdentity = Depends(resolve_operator_role)):
    """驳回审批单。被驳回的指纹再次触发不会重复建单前需先处理 pending（仅 admin）。"""
    _require_admin(operator)
    rec = decide_request(request_id, approve=False,
                         reviewer=_resolve_reviewer(request, body, operator), reason=body.reason)
    if rec is None:
        raise HTTPException(404, "审批单不存在或已处理")
    return {"ok": True, "request": rec}

File: frontend-admin/src/api/approvals.ts
/**
 * 工具审批 service
 *
 * 后端为 backend/app/api/routes/approvals.py（前缀 /approvals，
 * 经 next.config.js rewrite 由 /api/approvals 代理）。
 * reviewer 缺省时不传——后端取网关注入的 X-User-Name / X-User-Id
 * （P3 修复：审计字段不允许默认身份），都拿不到记 "unknown"。
 */

import { request } from '@/lib/fetcher'

const BASE = '/api/approvals'

export type ApprovalStatus = 'pending' | 'approved' | 'rejected' | 'executed'

/** 审批单（字段与后端 tool_approval store 的 record 对齐） */
export interface ApprovalRequest {
  id: string
  tool_name: string
  action: string
  /** 操作指纹（同指纹 pending 去重键），展示用可不关心 */
  fingerprint: string
  /** 工具调用参数（后端 JSON，结构随工具而定） */
  detail: Record<string, unknown>
  user_id: string
  status: ApprovalStatus
  reviewer: string | null
  reason: string | null
  created_at: string
  decided_at: string | null
  executed_at: string | null
}

export interface ApprovalListResponse {
  items: ApprovalRequest[]
  total: number
}

export interface ApprovalDecisionPayload {
  /** 审批理由：前端强制必填（写操作审计红线） */
  reason: string
  /** 不传则由后端取网关身份头 */
  reviewer?: string
}

export const approvalService = {
  list(status: ApprovalStatus | '', limit = 50): Promise<ApprovalListResponse> {
    const qs = new URLSearchParams({ limit: String(limit) })
    if (status) qs.set('status', status)
    return request<ApprovalListResponse>(`${BASE}?${qs.toString()}`)
  },

  approve(requestId: string, payload: ApprovalDecisionPayload): Promise<{ ok: boolean; request: ApprovalRequest }> {
    return request(`${BASE}/${encodeURIComponent(requestId)}/approve`, {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },

  reject(requestId: string, payload: ApprovalDecisionPayload): Promise<{ ok: boolean; request: ApprovalRequest }> {
    return request(`${BASE}/${encodeURIComponent(requestId)}/reject`, {
      method: 'POST',
      body: JSON.stringify(payload),
    })
  },
}
